contracts/provider/assembly/index.ts

```diff
@@ -0,0 +1,216 @@
+import {
+  u128,
+  storage,
+  context,
+  ContractPromiseBatch,
+  PersistentUnorderedMap,
+  logging,
+} from "near-sdk-as";
+import { onlyOwner } from "../../ownable/assembly";
+export {
+  transferOwnership,
+  renounceOwnership,
+  owner,
+} from "../../ownable/assembly";
+
+export const DefaultDivisor = u128.from("416000000000"); // yocto-Near per nano-second
+
+export const deposits = new PersistentUnorderedMap<string, Deposit>("d");
+
+@nearBindgen
+export class Deposit {
+  constructor(
+    public timestamp: u64,
+    public depositor: string,
+    public value: u128
+  ) {}
+}
+
+/**
+ * Proportion of deposited funds that will be kept by this provider upon release.
+ */
+export function providerProportion(): f64 {
+  return _providerProportion();
+}
+
+function _providerProportion(): f64 {
+  const p = storage.getString("_providerProportion");
+  if (p) return parseFloat(p);
+  return 0;
+}
+
+export function setProviderProportion(p: f64): void {
+  onlyOwner();
+  storage.set("_providerProportion", p.toString());
+}
+
+/**
+ * Divisor used to calculate session expiration.
+ *
+ * The session expiration is calculated as `timestamp + value / sessionDivisor`.
+ * Currently, the sessionDivisor is set to 40 milli-Ⓝ.
+ */
+export function sessionDivisor(): u128 {
+  return _sessionDivisor();
+}
+
+function _sessionDivisor(): u128 {
+  const d = storage.getString("_sessionDivisor");
+  if (d) return u128.fromString(d);
+  return u128.One;
+}
+
+export function setSessionDivisor(d: u128): void {
+  onlyOwner();
+  storage.set("_sessionDivisor", d.toString());
+}
+
+/**
+ * The off-chain upload/status RESTful API endpoint supported by this provider.
+ */
+export function apiEndpoint(): string {
+  return _apiEndpoint();
+}
+
+function _apiEndpoint(): string {
+  const p = storage.getString("_apiEndpoint");
+  return p ? p : "";
+}
+
+export function setApiEndpoint(a: string): void {
+  onlyOwner();
+  storage.setString("_apiEndpoint", a);
+}
+
+/**
+ * Logging event fired when a deposit has been added.
+ */
+function DepositAdded(
+  depositee: string,
+  depositor: string,
+  amount: u128
+): void {
+  logging.log(
+    `DepositAdded("${depositee}", "${depositor}", "${amount.toString()}")`
+  );
+}
+
+/**
+ * Logging event fired when a deposit has been released.
+ */
+function DepositReleased(
+  depositee: string,
+  depositor: string,
+  amount: u128
+): void {
+  logging.log(
+    `DepositReleased("${depositee}", "${depositor}", "${amount.toString()}")`
+  );
+}
+
+const KEY_INITIATED = "initiated";
+
+export function initialize(
```

Comment:
This is really just a helper function because NEAR doesn't have constructors. This is apparently "best practices". only the owner can do this, and they can only do it once.